people(followers): use `react-query` to fetch data & remove followers

flootr opened the pull request:
This PR is branched upon #50262 to make it easier to review. It should be rebased on `trunk` as #50262 gets in.

#### Changes proposed in this Pull Request

* People: migrate email-/followers to use `react-query`

#### Testing instructions

1. Go to the people section of a site and choose _Followers_
2. Make sure the list of followers shows up
3. If you scroll down on a page with more than 100 followers you should see another request with incremented `page` param
4. If you switch sites, make sure request params are reset and the list loads fine
5. Try to remove followers, are they deleted from the list and don't appear again on a reload?

Follow the steps above for _Email Followers_ as well.

flootr committed:
people(followers): use `react-query` to fetch data & remove followers

tyxla commented on client/data/followers/remove-follower.js:
```diff
@@ -0,0 +1,29 @@
+			return wp.undocumented().site( siteId )[ method ]( follower.ID );
+		},
+		{
+			onSuccess( data, variables ) {
```
Do we also need to handle the follower removal failure case somehow?

flootr replied:
Yes we should totally handle that case. Although not at this place. Error handling should be done where the custom hook is used depending on the returned `error` object.

```js
// function component body
const { removeFollower, error } = useRemoveFollower();

React.useEffect( () => {
  if ( error ) {
    doSomething();
  }
}, [ error ])
```

As far as I understand these `onSuccess`, `onError` are mainly used to handle hook internal state or data (like optimistic updates, etc).

I'm still thinking about a good solution to error handling in this specific case (there are a few gotchas which are specific to Calypso). Would you mind if I follow up with another PR? Error handling for removing followers seems to be lacking on prod as well so that'd be a great improvement.

tyxla replied:
Makes sense, I don't mind with following up in another PR at all 👍

tyxla reviewed (approved):
This is working great in my testing. Left a minor nit only. Super happy to see so much less boilerplate 👏

flootr committed:
useFollowers: expect `siteId` as first param

flootr committed:
followers: move `listKeyUp` & don't pass on `query`

flootr commented:
@tyxla based on the feedback I received on #50712 I made a few additions in https://github.com/Automattic/wp-calypso/pull/50378/commits/8c815422dc62a4388b3880268357300639bea3ac and https://github.com/Automattic/wp-calypso/pull/50378/commits/dd7a795525e3b8caf5b98c05691b72ecd48f77b3. Please let me know if those are fine.

flootr committed:
rename `useFollowers` to `useFollowersQuery`

flootr committed:
rename `useRemoveFollower` to `useRemoveFollowerMutation`

jsnajdr commented on client/data/followers/use-followers.js:
```diff
@@ -0,0 +1,59 @@
+	return useInfiniteQuery(
+		[ 'followers', siteId, type, search ],
+		async ( { pageParam = 1 } ) => {
+			const res = await wpcom.site( siteId ).statsFollowers( { ...fetchOptions, page: pageParam } );
```
It's interesting that the `.statsFollowers()` method calls the `/sites/:site/stats/followers` endpoint, while the original data layer handler calls `/sites/:site/followers`.

They both return exactly the same results. The only difference is that the `stats` one does an additional check testing if the Stats module is enabled on a Jetpack site. On Simple sites there's no difference at all.

For this UI that's outside Stats the `/followers` endpoint looks like the better one.

flootr replied:
Good point! Was about to change it anyway to use `wpcom.req.get` similar to https://github.com/Automattic/wp-calypso/pull/50712#discussion_r587311182.

jsnajdr commented on client/data/followers/use-followers.js:
```diff
@@ -0,0 +1,59 @@
+const compareUnique = ( a, b ) => a.ID === b.ID;
+
+const useFollowers = ( siteId, fetchOptions = {}, queryOptions = {} ) => {
+	const { search, type } = fetchOptions;
```
`type` should be a mandatory second arg that defaults to `wpcom`. If the `type` query arg is omitted from query, the REST endpoint will treat it as `wpcom`. And because `type` is used as part of the query key, we should avoid storing two identical responses under two keys, one with `type = wpcom` and other with `type = null`.

flootr replied:
`type` is set to `wpcom` by default in the `<FollowersList  />` component. Do you think we should move the default to `useFollowersQuery` instead?

flootr replied:
🤔 💭  If we make `type` a mandatory second argument (i.e. you can't omit it), does it make sense to set a default value in `useFollowersQuery` at all?

jsnajdr replied:
Yes, because `type` is not only used as parameter of `useFollowersQuery`, but also at many other places. E.g., UI labels.

jsnajdr commented on client/data/followers/use-followers.js:
```diff
@@ -0,0 +1,59 @@
+					...data,
+				};
+			},
+			...queryOptions,
```
Is it ever useful to override `getNextPageParam` or `select` with `queryOptions`? If not, the object should be constructed with the `{ ...queryOptions, getNextPageParam, select }` order.

flootr replied:
Yup, let's do it like you described 👍🏻 I can't think of a use-case right now and we can always change it back later.

jsnajdr commented on client/my-sites/people/followers-list/index.jsx:
```diff
@@ -3,41 +3,80 @@
+	} = useFollowers( site.ID, fetchOptions );
+	const { removeFollower } = useRemoveFollower();
+
+	useEffect( () => {
```
Let's also extract the `useErrorNotice` hook here.

jsnajdr commented on client/data/followers/remove-follower.js:
```diff
@@ -0,0 +1,29 @@
+function useRemoveFollower() {
+	const queryClient = useQueryClient();
+	const { mutate: removeFollower, ...rest } = useMutation(
+		( { siteId, follower, type } ) => {
```
All three arguments are mandatory so they could be three arguments:
```js
removeFollower( siteId, type, followerId )
```
The order is from the least specific (site) to the most specific (individual follower ID).

And we only need to pass the `follower.ID`, not the entire object.

flootr replied:
The `mutate` function returned from `useMutation` expects variables to be passed as object (see [reference](https://react-query.tanstack.com/reference/useMutation)) so we can't pass them as suggested. We could ofc use a custom signature for `removeFollower` which then wraps the original `mutate` fn instead of renaming it as is now.

jsnajdr replied:
I think a custom signature for `removeFollower` is fine. It denotes that some variables are mandatory.

`useMutation` return value also includes a promise-returning `mutateAsync`. If we're removing `mutate`, we should remove `mutateAsync`, too.

flootr replied:
Agreed. Either way I think we should not hide `mutate` and `mutateAsync` to make it possible for consumers to use the underlying "low level" method and don't introduce exceptions for certain returned properties.

flootr replied:
What do you think of https://github.com/Automattic/wp-calypso/pull/50378/commits/9395f5d3a20e8c493d64037da738b1b063c4eeaa?

jsnajdr replied:
Yes, just adding `removeFollower` to the object without removing anything sounds good. We'll see in future what parts of the `useMutation` return value we really use.

flootr committed:
useFollowersQuery: use `wpcom.req.get` directly

flootr committed:
FollowersList: extract `useErrorNotice`

flootr committed:
useFollowersQuery: make `type` mandatory & introduce `defaults`

flootr committed:
make `removeFollower` a wrapper around `mutate`

jsnajdr commented on client/data/followers/use-remove-follower-mutation.js:
```diff
@@ -0,0 +1,40 @@
+				followerId,
+			} ),
+		[ mutationObj ]
+	);
```
This `useCallback` won't optimize anything because `mutationObj` is different on each call. Using `mutationObj.mutate` as dependency is better.

flootr replied:
Addressed in https://github.com/Automattic/wp-calypso/pull/50378/commits/7c9aaf82fda1a8d5593759329dcfd7f88444f5f1

jsnajdr commented:
> does it make sense to set a default value in `useFollowersQuery` at all?

You'll then be able to call `useFollowersQuery( siteId )`, without type, and it will do something: load the `wpcom` list and store it under the `[ 'followers', siteId, undefined ]` key. And nobody will warn you.

Unless it's TypeScript, the second argument is never mandatory -- I can always call the function without any arguments.

flootr committed:
useFollowersQuery: set `type` to `wpcom` by default

flootr committed:
useFollowersQuery: change query params order

flootr committed:
removeFollower: use `mutate` as dep explicitly

flootr commented:
@jsnajdr I addressed your feedback and I think this PR is ready for another review 🙏🏻 
/cc @tyxla because this PR changed quite a bit since you reviewed it 😬

tyxla commented on client/data/followers/use-followers-query.js:
```diff
@@ -0,0 +1,65 @@
+ */
+import wpcom from 'calypso/lib/wp';
+
+const MAX_FOLLOWERS = 10; // means pages (= 1000 followers);
```
This feels a bit confusing. Can we rename it so it makes sense without the comment? `MAX_PAGES` or something maybe?

tyxla commented on client/data/followers/use-followers-query.js:
```diff
@@ -0,0 +1,65 @@
+			},
+			select: ( data ) => {
+				return {
+					/* @TODO: `uniqueBy` is necessary, because the API can return duplicates */
```
Nit: it's not really clear what we have to follow-up with from this comment. Could we expand / rewrite?

flootr replied:
That's true. `@TODO` is probably misleading here as it's not meant as something we have to follow up with in the client. Duplicates seem to be an existing problem which has to be resolved on the API side of things.

jsnajdr replied:
Do we know any details about this duplication? Is it just that different pages can have overlap? E.g., when I fetch page 1, then another client deletes some user on page 1, and then I fetch page 2? The entire list has shifted after the deletion. Or does the API genuinely return one user multiple times for some reason?

flootr replied:
> Or does the API genuinely return one user multiple times for some reason?

This seems to be the case here for some sites. More context here p1615806325122800-slack-C7YPUHBB2

tyxla replied:
As discussed there, it would be great if we can fix this problem in its roots - in other words, in the corresponding API endpoint.

tyxla commented on client/data/followers/use-followers-query.js:
```diff
@@ -0,0 +1,65 @@
+				}
+				return allPages.length + 1;
+			},
+			select: ( data ) => {
```
Could any special error handling be useful here?

flootr replied:
Can you elaborate on what you mean with special error handling?

tyxla replied:
Any error handling I guess 😉  But I think [your answer here](https://github.com/Automattic/wp-calypso/pull/50378#discussion_r593096012) explained that well already.

tyxla commented on client/data/followers/use-followers-query.js:
```diff
@@ -0,0 +1,65 @@
+				return {
+					/* @TODO: `uniqueBy` is necessary, because the API can return duplicates */
+					followers: uniqueBy( extractPages( data.pages ), compareUnique ),
+					total: data.pages[ 0 ].total,
```
Any chance this property may not exist for some reason?

flootr replied:
According to the [docs](https://react-query.tanstack.com/reference/useInfiniteQuery) `data.pages` should always be there. Even in an empty API response the `total` attribute should be defined as well. So in this case we should be safe. It shouldn't hurt though to change it to something like

```js
data.pages[ 0 ]?.total ?? 0 ;
```

This is what an empty response from the API looks like:

<img width="454" alt="Screenshot 2021-03-12 at 12 24 32" src="https://user-images.githubusercontent.com/9202899/110934218-0628bf00-832e-11eb-96d1-4b10632cb743.png">

tyxla replied:
Just took a look at the endpoint itself; seems like we'll always return a `total` if this was a successful request. So, feel free to leave it as-is.

tyxla commented on client/data/followers/use-followers-query.js:
```diff
@@ -0,0 +1,65 @@
+					/* @TODO: `uniqueBy` is necessary, because the API can return duplicates */
+					followers: uniqueBy( extractPages( data.pages ), compareUnique ),
+					total: data.pages[ 0 ].total,
+					...data,
```
Is all of `data` actually needed? Seems like we'll duplicate some of it, coming from `data.pages`, because we're already assigning that to additional propertiess.

flootr replied:
My intention was not to hide anything which is "natively" returned by the [`useInfiniteQuery`](https://react-query.tanstack.com/reference/useInfiniteQuery) hook to ensure we're not hiding any documented props from consumers. It's debatable whether that approach is how we want to use custom query hooks and probably part of us developing conventions using ReactQuery.

/cc @jsnajdr might have an opinion here as well.

flootr replied:
By the way I found [this note in the ReactQuery](https://react-query.tanstack.com/guides/infinite-queries) docs which says

> When using options like `initialData` or `select` in your query, make sure when you restructure your data that it still includes `data.pages` and `data.pageParams` properties, otherwise your changes will be overwritten by the query in its return

jsnajdr replied:
I don't know yet what is going to be the ideal API for hooks built on `useQuery` or `useMutation`. Should we return only some selected data and hide everything else from the low-level hooks? Or should we only add new properties and expose the underlying `react-query` API fully? We'll see as we start using the library at more places.

tyxla commented on client/data/followers/use-remove-follower-mutation.js:
```diff
@@ -0,0 +1,40 @@
+			return wp.undocumented().site( siteId )[ method ]( followerId );
+		},
+		{
+			onSuccess( data, variables ) {
```
Missing error handling?

flootr replied:
Generally errors should be handled where the `useRemoveFollowerMutation` hook is used, not hook-internally. As far as I understand those `onX` hooks are used to manipulate data and state internally e.g. optimistic updates. Adding error handling here would add it for any usage of the which might not be what we want.

tyxla commented on client/my-sites/people/followers-list/followers.jsx:
```diff
@@ -46,13 +36,17 @@ class Followers extends Component {
-		this.props.incrementPage();
-		gaRecordEvent( 'People', analyticsAction, 'page', this.props.currentPage + 1 );
+		this.props.fetchNextPage();
+		this.props.recordGoogleEvent( 'People', analyticsAction, 'page', this.props.currentPage + 1 );
```
The `gaRecordEvent` -> `recordGoogleEvent` refactor seems like a perfect candidate for a separate PR, if you wanted to split this one into smaller pieces.

flootr replied:
True 👍🏻 I'll split that part into another PR.

flootr replied:
Extracted here: #51061

tyxla commented on client/my-sites/people/followers-list/followers.jsx:
```diff
@@ -96,16 +90,13 @@ class Followers extends Component {
-			! this.props.fetching
-		);
+		const { search, followers, isFetching } = this.props;
+		return search && ! followers.length && ! isFetching;
```
Could we reuse `this.siteHasNoFollowers()` here?

```suggestion
		return search && this.siteHasNoFollowers();
```

flootr replied:
done via https://github.com/Automattic/wp-calypso/pull/50378/commits/959dab9b048121f9564d0c809eecffee6fd4255b – I had to remove unused properties one line above which is why I didn't apply that suggestion.

tyxla commented on client/my-sites/people/followers-list/index.jsx:
```diff
@@ -3,41 +3,80 @@
+			dispatch(
+				errorNotice( notice, {
+					id: 'site-followers-notice',
+					button: 'Try again',
```
Should we localize this string?

flootr replied:
done via https://github.com/Automattic/wp-calypso/pull/50378/commits/1d351dc99e06a4b71b3a5070cf96283a41eaabd7

jsnajdr commented on client/data/followers/use-followers-query.js:
```diff
@@ -0,0 +1,65 @@
+ * @returns {object}          A normalized follower object
+ */
+export function normalizeFollower( follower ) {
+	follower.avatar_URL = follower.avatar;
```
This assignment should not mutate the `follower` object. We're mutating an object that is in React Query cache and the `avatar_URL` field then will be also on the supposedly raw `data.pages[ x ].subscribers[ y ]` response.

flootr replied:
Good catch @jsnajdr, will add a fix 👍🏻

flootr replied:
Would https://github.com/Automattic/wp-calypso/pull/50378/commits/ce86cace8bbfc5166ccd67b2625debfca73164b2 work as a fix?

flootr committed:
FollowersList: make sure strings are translated

flootr committed:
FollowersList: simplify `noFollowerSearchResult`

flootr committed:
useFollowersQuery: don't mutate internal data when normalizing

jsnajdr commented on client/data/followers/use-remove-follower-mutation.js:
```diff
@@ -0,0 +1,40 @@
+		[ mutate ]
+	);
+
+	return { removeFollower, mutate, ...mutationObj };
```
Removing the `mutate` prop by `...mutationObj` destructuring and then adding it again ... you didn't need to do that. I'd personally write it as:
```js
const mutation = useMutation();
const removeFollower = useCallback( () => mutation.mutate(), [ mutation.mutate ] );
return { ...mutation, removeFollower };
```

flootr replied:
That doesn't seem to work properly. `useCallback` complains about missing dependencies when using it that way.

<img width="1490" alt="Screenshot 2021-03-16 at 09 08 12" src="https://user-images.githubusercontent.com/9202899/111276494-ac7b0a00-8637-11eb-89b5-4e1594e7ecde.png">

jsnajdr replied:
I see, that's a deficiency of the ESLint rule -- at runtime it would work reliably.

In that case I would do:
```js
const mutation = useMutation();
const { mutate } = mutation;
...
```
But this all is just my personal style preference. I'd try to preserve the `mutation` object and only add new fields to it. Don't feel obliged to follow it 🙂

flootr replied:
I think that's a great suggestion 👍🏻

flootr committed:
useRemoveFollowerMutation: change how `mutate` is extracted from `useMutation`

jsnajdr reviewed (approved):
:shipit:
